File: day01/main.cpp
// https://adventofcode.com/2024/day/1

#include <algorithm>
#include <fstream>
#include <iostream>
#include <map>
#include <sstream>
#include <vector>

int main() {
  // part 1
  std::vector<int> left;
  std::vector<int> right;
  std::vector<int> distances;
  std::ifstream inputFile("input.txt");
  std::string line;

  // line contains two numbers separated by a space
  while (std::getline(inputFile, line)) {
    std::istringstream iss(line);
    int n1, n2;
    iss >> n1 >> n2;
    left.push_back(n1);
    right.push_back(n2);
  }

  std::sort(left.begin(), left.end());
  std::sort(right.begin(), right.end());

  int sum = 0;
  for (int i = 0; i < left.size(); i++) {
    int distance = left[i] - right[i];
    if (distance < 0) {
      distance = -distance;
    }
    sum += distance;
  }

  std::cout << "Part 1: " << sum << std::endl;

  // Part 2
  std::map<int, int> appearsOnRight;

  for (const int l : left) {
    int count = 0;
    for (const int r : right) {
      if (r == l)
        count++;
    }
    appearsOnRight[l] = count;
  }

  int sum2 = 0;
  for (const int l : left) {
    sum2 += l * appearsOnRight[l];
  }

  std::cout << "Part 2: " << sum2 << std::endl;

  return 0;
}

File: day01/src/main.rs
use std::collections::HashMap;
use utility::{get_raw_text, get_vec_from_txt};

fn main() {
    let txt = get_raw_text("input.txt").unwrap();
    let mut left = Vec::new();
    let mut right = Vec::new();

    for line in get_vec_from_txt(&txt) {
        for (i, elem) in line.split_whitespace().enumerate() {
            let n: i64 = elem.parse().expect("All elements should be parsable");
            match i {
                0 => left.push(n),
                1 => right.push(n),
                _ => unreachable!("There should be only two values per line"),
            }
        }
    }

    left.sort();
    right.sort();

    let mut sum = 0;
    for i in 0..left.len() {
        let distance = left[i].abs_diff(right[i]);
        sum += distance;
    }

    println!("Part 1: {}", sum);

    // Part 2
    let mut appears_on_right = HashMap::new();

    for r in &right {
        *appears_on_right.entry(r).or_insert(0) += 1;
    }

    let sum2: i64 = left
        .iter()
        .filter_map(|l| appears_on_right.get(l).map(|&count| l * count))
        .sum();

    println!("Part 2: {}", sum2);
}
